restapi/src/test/java/io/stargate/web/docsapi/service/query/rules/ExpressionUtilsTest.java

```diff
@@ -0,0 +1,199 @@
+import io.stargate.web.docsapi.service.query.filter.operation.FilterOperationCode;
+import io.stargate.web.docsapi.service.query.filter.operation.ValueFilterOperation;
+import io.stargate.web.docsapi.service.query.filter.operation.impl.GtFilterOperation;
+import io.stargate.web.docsapi.service.query.filter.operation.impl.LteFilterOperation;
+import java.util.Collections;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+class ExpressionUtilsTest {
+
+  private FilterExpression e(FilterOperationCode op, int value) {
+    FilterPath filterPath = ImmutableFilterPath.of(Collections.singletonList("field"));
+
+    ValueFilterOperation operation;
+    switch (op) {
+      case GT:
+        operation = GtFilterOperation.of();
+        break;
+      case LTE:
+        operation = LteFilterOperation.of();
+        break;
+      default:
+        throw new UnsupportedOperationException("Unsupported op code: " + op);
+    }
+
+    BaseCondition condition = ImmutableNumberCondition.of(operation, value);
+    return ImmutableFilterExpression.of(filterPath, condition, 0);
+  }
+
+  @Nested
+  class CNF {
+
+    @Test
+    void oneCondition() {
+      Expression<FilterExpression> cnf = toSimplifiedCnf(e(GT, 1));
+      assertThat(cnf.toLexicographicString()).isEqualTo("field GT 1");
+    }
+
+    @Test
+    void simpleAnd() {
+      Expression<FilterExpression> cnf = toSimplifiedCnf(And.of(e(GT, 1), e(LTE, 2)));
+      assertThat(cnf.toLexicographicString()).isEqualTo("(field GT 1 & field LTE 2)");
+    }
+
+    @Test
+    void simpleOr() {
+      Expression<FilterExpression> cnf = toSimplifiedCnf(Or.of(e(GT, 1), e(LTE, 2)));
+      assertThat(cnf.toLexicographicString()).isEqualTo("(field GT 1 | field LTE 2)");
+    }
+
+    @Test
+    void orOfAnd() {
+      Expression<FilterExpression> cnf =
+          toSimplifiedCnf(Or.of(e(LTE, 1), And.of(e(GT, 2), e(LTE, 3))));
+      assertThat(cnf.toLexicographicString())
+          .isEqualTo("((field GT 2 | field LTE 1) & (field LTE 1 | field LTE 3))");
+    }
+
+    @Test
+    void andOfOr() {
+      Expression<FilterExpression> cnf =
+          toSimplifiedCnf(And.of(e(LTE, 1), Or.of(e(GT, 2), e(LTE, 3))));
+      assertThat(cnf.toLexicographicString())
+          .isEqualTo("((field GT 2 | field LTE 3) & field LTE 1)");
+    }
+
+    @Test
+    void wrappingOr() {
+      Expression<FilterExpression> cnf =
+          toSimplifiedCnf(Or.of(And.of(e(LTE, 1), Or.of(e(GT, 2), e(LTE, 3)))));
+      assertThat(cnf.toLexicographicString())
+          .isEqualTo("((field GT 2 | field LTE 3) & field LTE 1)");
+    }
+
+    @Test
+    void orOfAndWithNegation() {
+      Expression<FilterExpression> cnf =
+          toSimplifiedCnf(Or.of(Not.of(e(LTE, 1)), e(GT, 2), And.of(e(GT, 3), e(LTE, 4))));
+      // Note: the first predicate got inverted
+      // Note: predicate-specific simplification is not supported (yet?), as in:
```

Comment:
let's add a ticket for this.. should be supper easy to add a rule for this..

Reply:
#1165